Blog post on UML and the Strategy Design Pattern (#52)
ImperialCollegeLondon/RSEBlog: docs/posts/design_patterns.md

---
date: 2024-08-30
author:
  - HarmonicReflux
categories:
  - Technology
tags:

>   - Design Patterns / Gang of Four
>   - Object Oriented Programming (OOP) / Functional Programming
>   - Python
>   - Financial Data Science
>   - Software Engineering principles

Reviewer: I would suggest simplifying the tags as the following, considering you don't touch functional programming:
```suggestion
  - Design Patterns
  - Object Oriented Programming
  - Python
  - Financial Data Science
  - Software Engineering principles
  - UML Diagrams
```

Author: I made these changes in my local copy.

Reviewer: And why are not in the remote?

---

> # Applying the Strategy Pattern: UML diagrams and Python for European option pricing

Reviewer 2: This title seems a bit difficult to unpack, given the intended audience might not yet know about any of them. Maybe something less specific like e.g.  'Code structuring with UML diagrams for financial software'.

The book referenced above is approximately 30 years old, and much has changed in the computing world since its publication, including the evolution of programming languages and development methodologies.
The 'move fast and fail fast' mentality of start-ups, and the approach often seen in quantitative analysis within finance—where the focus is primarily on solving the immediate problem rather than on
the design of the program—are prime examples of this shift. Many Jupyter notebooks, written in isolation, exemplify this mindset, where data scientists are often stuck on the
'make it work' mantra, without progressing to 'make it right,' and ultimately, 'make it fast.'

This is not meant as a criticism, as the 'make it work' step often involves significant reworking, rewriting, or even re-manufacturing of entire parts of the code.
Over-engineering at the prototyping or proof-of-concept stage can be counterproductive. However, as systems mature, this approach can lead to technical debt,
resulting in code that is difficult, if not impossible, to maintain.

> In the following, we first describe UML diagrams and how they are used to plan and visualise software architecture before coding itself.

Reviewer: This paragraph here comes a bit out of the blue after the nice introduction above. I would suggest something along the lines of:

> In this blog post we will... (some indication of the overall purpose of the post, linking the above introduction to what is coming next). We first start describing UML diagrams...

Reviewer 2: Agree. Describing how this blog post will address the issues identified in the good introduction above would make the contents and goal of the post more clear. Even if it's obvious to you, it's a reassurance for the reader that they are not just being clickbaited.

Author: It is the least of my intention to deliver click-bait. I have a few concepts to explain here: 
- The design patterns and its purpose
- How it can be applied to a concrete example (here: option pricing from finance which essentially is pricing an insurance contract given uncertainty of future evolution of an underlying asset (and hence it is called a "derivative"))
- Incidentally, this forces me to also explain the concept of said derivatives as otherwise it pops out of nowhere for readers unfamiliar with the topic. 

Hence, it is not that easy to cater for all interests in the introduction.

This is then followed by the definition and implementation of the Strategy pattern. We conclude this article with a practical implementation of pricing European options, ending with a discussion.

<!-- more -->

## Unified Modeling Language (UML)

Before diving into Design Patterns, it is important to first consider how we can define objects and their properties at the conceptual level, much like an architect creates a
blueprint before actual construction begins.
In software architecture, a key tool for this purpose is the Unified Modeling Language (UML).
UML is a standardised modeling language widely used in software engineering to specify, visualise, construct, and document the components of software systems.

> While it is not a programming language, UML serves as a helpful tool for helping software developers plan and design software architecture.

Reviewer 2: ... and also helps in the onboarding of new developers in the team.

### How does UML help us in Software Engineering?

> UML includes a set of graphic notation techniques to create visual representations of Object-Oriented software systems.

Reviewer: So, UML diagrams are not useful for functional programming?

Author: No, UML's are universal in nature, as the name suggests.

> UML simplifies communication, making it easier for both development teams and external stakeholders to understand the software design.

Reviewer 2: ... also incentivises forward planning of a structured API, rather than retroactively having to standardise a crap API in the future.

Author: Great. I added that to the post.

Another key strength of UML is its platform independence, allowing the same modeling language to be used across various platforms, ensuring consistency and flexibility in software design.

### Types of UML diagrams

UML includes several types of diagrams, each serving different purposes:

- Structural diagrams (e.g., class diagrams, object diagrams): Focus on the static aspects of the system.

- Behavioural diagrams (e.g., use case diagrams, activity diagrams): Focus on dynamic aspects.

- Interaction diagrams (part of behavioral diagrams, e.g., sequence diagrams, communication diagrams): Focus specifically on the flow of control and data among the elements in the system.

Thus, UML helps developers engineer and translate their ideas into functional software, serving as a component of modern software engineering practice.

### UML within Jupyter notebooks

From within Jupyter notebooks, one can design UML diagrams using a library called `iplantuml`, that we can install via pip upon invoking `pip install iplantuml`.

```python
try:
    import iplantuml
except ImportError:
    !pip install iplantuml
    import iplantuml
```

> ## The `Strategy` Design Pattern

Reviewer: You are talking about UML and then suddenly jump to describe specific design patterns. There has to be some link between both sections for the post to read well and be fluid. Why UML helps you with this design pattern? I'm sure you know it, but it is not clear in the text.

Reviewer 2: Agree.

Author: The idea is to have a visual representation, or a representation at all for that matter, that tells the reader what is being implemented. Seeing the code as such and translating it to a "structure" is difficult. Seeing the "structure" beforehand, i.e. metaphorically speaking, the blue print, before the actual building is built should make it easier.

This pattern allows the algorithm to vary independently of the clients that use it.

The concept of the Strategy Pattern can be outlined as follows.

- Context: This is the class that uses a Strategy. The Context is configured with a ConcreteStrategy object, which it uses to perform a specific algorithm.
- Strategy interface: This defines a common interface for all supported algorithms. The context uses this interface to call the algorithm defined by a ConcreteStrategy.
- ConcreteStrategy: These are the classes that implement the Strategy interface, each providing a specific implementation of the algorithm.

We can code up the UML diagram of the Strategy Pattern as follows.

```plantuml
%%plantuml
@startuml
class Context {
}

interface Strategy {
    +execute()
}

class ConcreteStrategy1 {
    +execute()
}

class ConcreteStrategy2 {
    +execute()
}

Strategy <|.. ConcreteStrategy1
Strategy <|.. ConcreteStrategy2
Context "1" *-- "*" Strategy
@enduml
```

![Definition of the Strategy Pattern](images/design_patterns/def_strategy_pattern.png)
<!-- markdownlint-disable-next-line MD036 -->
*The rendered result should look similar to the above figure.*

> 

Reviewer 2: Here would be a perfect point to link back to the issues described in the intro: how does using the strategy pattern (also planning and documenting it with the nice UML diagram) help alleviate the intro issues? It's obvious to us, of course, but presumably the person reading this is unfamiliar with softeng norms and doesn't see the problem with churning out crap code fast.

Author: I addressed the issue you point out here.

We can now use this our blueprint and engineer software that implements the abstract pattern.

> As an example we pick pricing a European call and put option using the `Black-Scholes-Merton` equation, which is

Reviewer 2: I'd hazard a guess that 80% of the audience doesn't know what, or the significance of, any of these things. You may need to spoon feed people a bit here. Briefly why use/what is a European call, the Black-Scholes-Merton equation. Some people might not even know what a PDE is.

Author: I did address this point in my latest commit.

- Association:
`Main "1" *-- "1" BlackScholesOption` indicates an association relationship between `Main` and `BlackScholesOption`. The `1` near both `Main` and `BlackScholesOption` signifies a one-to-one relationship.
Each instance of Main is associated with exactly one instance of `BlackScholesOption`. The filled diamond indicates a composition relationship where
`Main` strongly owns or controls the lifecycle of the `BlackScholesOption`.

- Code context:
`OptionPriceStrategy` interface defines a contract for computing option prices.
It has one method, `compute_price`, which takes an instance of `BlackScholesOption` and returns a float.

- `EuropeanCallStrategy` and `EuropeanPutStrategy Classes` implement the `OptionPriceStrategy` interface.
Each provides a specific implementation of the `compute_price` method for `European call and put` options, respectively.

- `BlackScholesOption class` contains attributes for the option parameters (e.g., stock price, strike price, time to maturity, risk-free rate, volatility).
It has a (private) method `_compute_d1_d2` to calculate the `d1` and `d2` values used in the Black-Scholes formula.

> The `_compute_option_price` method takes an `OptionPriceStrategy` instance and uses it to compute the option price.

Reviewer 2: I'd make it clear here the link to the language used in the Strategy pattern description. "The BlackScholesOption class implements the 'context' of the pattern described earler... " or words to that effect.

- `Main Class`: Responsible for creating instances of `BlackScholesOption` and the strategy classes (`EuropeanCallStrategy and EuropeanPutStrategy`).
It then uses these to compute and print the option prices

### Python code

Translating the UML diagram to Python code can look like so:

```python
from abc import ABC, abstractmethod
from dataclasses import dataclass
import numpy as np
from scipy.stats import norm
from typing import Tuple

# define the Strategy interface

> class OptionPriceStrategy(ABC):
>     @abstractmethod
>     def compute_price(self, option: 'BlackScholesOption') -> float:
>         pass
> 
> # implement the European call option pricing strategy
> class EuropeanCallStrategy(OptionPriceStrategy):
>     def compute_price(self, option: 'BlackScholesOption') -> float:
>         d1, d2 = option.compute_d1_d2()
>         price = (option.stock_price * norm.cdf(d1) -
>                  option.strike_price * np.exp(-option.risk_free_rate * option.time_to_maturity) * norm.cdf(d2))
>         return price
> 
> # implement the European put option pricing strategy
> class EuropeanPutStrategy(OptionPriceStrategy):
>     def compute_price(self, option: 'BlackScholesOption') -> float:
>         d1, d2 = option.compute_d1_d2()
>         price = (option.strike_price * np.exp(-option.risk_free_rate * option.time_to_maturity) * norm.cdf(-d2) -
>                  option.stock_price * norm.cdf(-d1))
>         return price

Reviewer: I don't understand why these need to be classes. That IS overengineer the problem. The following will work equally well, and keep any static type checker happy:

```suggestion
OptionPriceStrategy: Callable[['BlackScholesOption'], float]

# implement the European call option pricing strategy
def european_call_strategy(self, option: 'BlackScholesOption') -> float:
    d1, d2 = option.compute_d1_d2()
    price = (option.stock_price * norm.cdf(d1) -
        option.strike_price * np.exp(-option.risk_free_rate * option.time_to_maturity) * norm.cdf(d2))
    return price

# implement the European put option pricing strategy
def european_put_strategy(self, option: 'BlackScholesOption') -> float:
    d1, d2 = option.compute_d1_d2()
    price = (option.strike_price * np.exp(-option.risk_free_rate * option.time_to_maturity) * norm.cdf(-d2) -
        option.stock_price * norm.cdf(-d1))
    return price
```

And then... see other comments.

Author: I was inspired by the original Gang of Four book and aimed at translating their scaffold into Python. I am aware a functional approach is often more direct to implement and maybe I am guilty of "disguising something as an object that just as well can be a function". I chose the OOP approach on purpose and it was the initial goal to code up the Strategy Pattern in an OOP fashion.

# define the context class, representing the Black-Scholes option
@dataclass
class BlackScholesOption:
    stock_price: float
    strike_price: float
    time_to_maturity: float
    risk_free_rate: float
    volatility: float

    # method to compute d1 and d2, which are intermediate calculations for option pricing
    def compute_d1_d2(self) -> Tuple[float, float]:
        d1 = (np.log(self.stock_price / self.strike_price) +
              (self.risk_free_rate + 0.5 * self.volatility ** 2) * self.time_to_maturity) / (self.volatility * np.sqrt(self.time_to_maturity))
        d2 = d1 - self.volatility * np.sqrt(self.time_to_maturity)
        return d1, d2

>     # method to compute the option price using the provided strategy
>     def compute_option_price(self, strategy: OptionPriceStrategy) -> float:
>         return strategy.compute_price(self)
> 
> # main function to demonstrate the usage of the Strategy Pattern
> def main() -> None:
>     # create instances of the pricing strategies
>     call_strategy: OptionPriceStrategy = EuropeanCallStrategy()
>     put_strategy: OptionPriceStrategy = EuropeanPutStrategy()
> 
>     # create a Black-Scholes option with given parameters
>     option: BlackScholesOption = BlackScholesOption(
>         stock_price=100,
>         strike_price=100,
>         time_to_maturity=1,
>         risk_free_rate=0.05,
>         volatility=0.2
>     )
> 
>     # compute and print the call option price using the call strategy
>     call_price: float = option.compute_option_price(call_strategy)
>     print(f"Call price: {call_price:.2f} [monetary units].")
> 
>     # compute and print the put option price using the put strategy
>     put_price: float = option.compute_option_price(put_strategy)
>     print(f"Put price: {put_price:.2f} [monetary units].")

Reviewer: And this can be re-written as:
```suggestion
    # method to compute the option price using the provided strategy
    def compute_option_price(self, strategy: OptionPriceStrategy) -> float:
        return strategy(self)

# main function to demonstrate the usage of the Strategy Pattern
def main() -> None:
    # create a Black-Scholes option with given parameters
    option: BlackScholesOption = BlackScholesOption(
        stock_price=100,
        strike_price=100,
        time_to_maturity=1,
        risk_free_rate=0.05,
        volatility=0.2
    )

    # compute and print the call option price using the call strategy
    call_price: float = option.compute_option_price(european_call_strategy)
    print(f"Call price: {call_price:.2f} [monetary units].")

    # compute and print the put option price using the put strategy
    put_price: float = option.compute_option_price(european_put_strategy)
    print(f"Put price: {put_price:.2f} [monetary units].")
```

In summary: don't make classes out of things that do not need to be classes.

Each pattern provides a tried-and-true solution to a specific problem.
Some of these patterns focus on object creation, others on structuring systems of objects, and still others on how objects should communicate.

The book referenced above is approximately 30 years old, and much has changed in the computing world since its publication,
including the evolution of programming languages and development methodologies.
The 'move fast and fail fast' mentality of start-ups, and the approach often seen in quantitative analysis within finance—where the focus
is primarily on solving the immediate problem rather than on
the design of the program—are prime examples of this shift. Many Jupyter notebooks, written in isolation, exemplify this mindset, where data scientists are
often stuck on the 'make it work' mantra, without progressing to 'make it right,' and ultimately, 'make it fast.'

This is not meant as a criticism, as the 'make it work' step often involves significant reworking, rewriting, or even re-manufacturing of entire parts of the code.
However, as systems mature, this approach can lead to technical debt, resulting in code that is confusingly difficult to maintain.

> This blog post illustrates how the concept of a `Design Pattern` from the early 90ies can be translated to modern day Python. Given the 23 patterns introduced

Reviewer 3: ```suggestion
This blog post illustrates how the concept of a `Design Pattern` from the early '90s can be translated to modern day Python. Given the 23 patterns introduced
```

Each pattern provides a tried-and-true solution to a specific problem.
Some of these patterns focus on object creation, others on structuring systems of objects, and still others on how objects should communicate.

The book referenced above is approximately 30 years old, and much has changed in the computing world since its publication,
including the evolution of programming languages and development methodologies.
The 'move fast and fail fast' mentality of start-ups, and the approach often seen in quantitative analysis within finance—where the focus
is primarily on solving the immediate problem rather than on
the design of the program—are prime examples of this shift. Many Jupyter notebooks, written in isolation, exemplify this mindset, where data scientists are
often stuck on the 'make it work' mantra, without progressing to 'make it right,' and ultimately, 'make it fast.'

This is not meant as a criticism, as the 'make it work' step often involves significant reworking, rewriting, or even re-manufacturing of entire parts of the code.
However, as systems mature, this approach can lead to technical debt, resulting in code that is confusingly difficult to maintain.

> This blog post illustrates how the concept of a `Design Pattern` from the early 90ies can be translated to modern day Python. Given the 23 patterns introduced

Reviewer 3: This is the source of one of the spelling errors ("ies"):

```suggestion
This blog post illustrates how the concept of a `Design Pattern` from the early '90s can be translated to modern day Python. Given the 23 patterns introduced
```

- Context: This is the class that uses a Strategy. The Context is configured with a `ConcreteStrategy` object, which it uses to perform a specific algorithm.
- Strategy interface: This defines a common interface for all supported algorithms. The context uses this interface to call the algorithm defined by a `ConcreteStrategy`.
- `ConcreteStrategy`: These are the classes that implement the Strategy interface, each providing a specific implementation of the algorithm.

We can code up the UML diagram of the Strategy pattern as follows.

```plantuml
%%plantuml
@startuml
class Context {
    +context_data: float
    +execute_strategy(strategy: Callable) : float
}

note right of Context : Represents data and the method to execute a strategy

interface Strategy {
    +__call__(data) : float
}

note right of Strategy : Callable functions used as strategies

strategy1 : float __call__(data)
strategy2 : float __call__(data)

Context --> Strategy
Context --> strategy1 : uses
Context --> strategy2 : uses
@enduml
```

![Definition of the Strategy Pattern](images/design_patterns/def_strategy_pattern.png)
<!-- markdownlint-disable-next-line MD036 -->

> *The rendered result should look similar to the above figure.*

Reviewer 3: For figure captions, you can do this:

```md
![My figure caption.](path/to/image.png)
```

Author: I follow your suggestion for the syntax, but am not able to break the caption like this

```mk
![This is the first line of the caption.<br>This should be the second line of the caption](path/to/image.png)
```

The result markdown renders doesn't look that great. Am thankful for suggestions.

> We can identify the following relations in the UML diagram

Reviewer 3: ```suggestion
We can identify the following relations in the UML diagram:
```

the `Strategy pattern`.

## Implementation of the Strategy pattern for European option pricing

We make use of the definition of the Strategy design pattern above and adapt it to the problem at hand.
We first modify the UML diagram to lay out the architecture and then follow-up
with a code implementation.

### UML diagram

```plantuml
%%plantuml
@startuml
class BlackScholesOption {
    +float stock_price
    +float strike_price
    +float time_to_maturity
    +float risk_free_rate
    +float volatility
    +compute_d1_d2() : Tuple[float, float]
    +compute_option_price(strategy: OptionPriceStrategy) : float
}

interface OptionPriceStrategy {
    +__call__(option: BlackScholesOption) : float
}

BlackScholesOption --> OptionPriceStrategy

european_call_strategy : float __call__(option: BlackScholesOption)
european_put_strategy : float __call__(option: BlackScholesOption)

BlackScholesOption --> european_call_strategy : uses
BlackScholesOption --> european_put_strategy : uses

@enduml
```

![Application of the Strategy Pattern for the pricing European call and put options](images/design_patterns/option_pricing_using_strategy_pattern.png)
<!-- markdownlint-disable-next-line MD036 -->

> *The rendered result should look similar to the above figure.*

Reviewer 3: See my comment above re figure captions.

        d2 = d1 - self.volatility * np.sqrt(self.time_to_maturity)
        return d1, d2

    def compute_option_price(self, strategy: OptionPriceStrategy) -> float:  # method to compute the option price using a given pricing strategy
        return strategy(self)

def main() -> None:

    option = BlackScholesOption(  # create a Black-Scholes option with given parameters
        stock_price=100,
        strike_price=100,
        time_to_maturity=1,
        risk_free_rate=0.05,
        volatility=0.2
    )

    call_price: float = option.compute_option_price(european_call_strategy)  # compute and print the call option price using the call strategy
    print(f"call price: {call_price:.2f} [monetary units].")

    put_price: float = option.compute_option_price(european_put_strategy)  # compute and print the put option price using the put strategy
    print(f"put price: {put_price:.2f} [monetary units].")

if __name__ == "__main__":
    main()
```

In the currency of the contract, under the Black-Scholes model, the price of a European call and
put options given the underlying stock are hence:

> ```zsh

Reviewer 3: This isn't a shell script :stuck_out_tongue:. Maybe this will work:

```suggestion
```txt
```

Author: Lovely. That's what I was looking for in the first place.

functional programming paradigm that Python supports.
To illustrate the practical use of this design pattern, we provided a concrete example from
finance, namely pricing European put and call options.
We created a dataclass that stores the data we need our strategies to take as input.

By applying the Strategy pattern, we achieved a clear separation of concerns, resulting in
loosely coupled and maintainable code. These are key characteristics of high-quality software
engineering, ensuring that the code is understandable, extendable, and maintainable.

Also, in an addendum, we made use of the computed output and utilised the put-call parity
to confirm the options are priced fair.

The reader is encouraged to explore the original publication on design patterns for
to see what other patterns exist and what problems they solve.
As the design patterns themselves are generally valid, feel free to implement them
programming language of your choice. Finally, if you are interested in a real-world data
science application, you could plug into an options exchange and
observe whether you can spot any arbitrage opportunities, grated this will be an ambitious project.

## References

> <!-- spell-check-ignore -->

Reviewer 3: If this works, then we can remove "Modeling" from `.wordlist.txt`, right?

Author: Yes, that should work. As with C++, I now have code that "compiles", and I now can work on make it more beautiful. But honestly, spellchecker has been a pain and to make it "behave" has been tricky. 

As for  the Latin phrase "per se", there is not much I can do about it. I do not use them that regularly, although they come across in economic language quite regularly, e.g. "ex-post [comparison]", "cum-ex [date]" , "ceteris paribus", and frankly e.g. and etc. are Latin words as well.   

I will remove "Modeling" from the wordlist. I cite the book as a reference. As it is by an American author, the title is with a single l rather a double l.

Reviewer 3: I know it's correct -- we don't want to be changing the spelling of a title. I just meant that the `<!-- spell-check-ignore -->` bit should mean that it isn't flagged.

Reviewer 3: Re the Latin words (hah -- re is another), I wasn't asking you to change your text! I was just wondering if we could fix the spellchecker for some of the things that don't look like words (e.g. "nd" which I guess is part of "2nd" or something).

Author: I also changed `Unified Modeling Language` to `Unified Modelling Language` to comply 
with the orthography of British English.
